packages/Search/src/details/DTK_DetailsPredicate.hpp

```diff
@@ -24,115 +24,58 @@ struct SpatialPredicateTag
 {
 };
 
-// COMMENT: Default constructor and assignment operator are required to be able
-// to declare a Kokkos::View of a predicate type and fill it with a
-// Kokkos::for_parallel.
-
+template <typename Geometry>
 struct Nearest
 {
     using Tag = NearestPredicateTag;
 
     KOKKOS_INLINE_FUNCTION
-    Nearest()
-        : _query_point( {{0., 0., 0.}} )
-        , _k( 0 )
-    {
-    }
-
-    KOKKOS_INLINE_FUNCTION Nearest &operator=( Nearest const &other )
-    {
-        _query_point = other._query_point;
-        _k = other._k;
-        return *this;
-    }
+    Nearest() = default;
 
     KOKKOS_INLINE_FUNCTION
-    Nearest( Point const &query_point, int k )
-        : _query_point( query_point )
+    Nearest( Geometry const &geometry, int k )
+        : _geometry( geometry )
         , _k( k )
     {
     }
 
-    Point _query_point;
-    int _k;
+    Geometry _geometry;
+    int _k = 0;
 };
 
-class Within
+template <typename Geometry>
+struct Intersects
 {
-  public:
     using Tag = SpatialPredicateTag;
 
-    KOKKOS_INLINE_FUNCTION
-    Within()
-        : _query_point( {{0., 0., 0.}} )
-        , _radius( 0. )
-    {
-    }
-
-    KOKKOS_INLINE_FUNCTION Within &operator=( Within const &other )
-    {
-        _query_point = other._query_point;
-        _radius = other._radius;
-        return *this;
-    }
+    KOKKOS_INLINE_FUNCTION Intersects() = default;
 
-    KOKKOS_INLINE_FUNCTION
-    Within( Point const &query_point, double const radius )
-        : _query_point( query_point )
-        , _radius( radius )
+    KOKKOS_INLINE_FUNCTION Intersects( Geometry const &geometry )
+        : _geometry( geometry )
     {
     }
 
     KOKKOS_INLINE_FUNCTION
     bool operator()( Node const *node ) const
     {
-        double node_distance = distance( _query_point, node->bounding_box );
-        return ( node_distance <= _radius ) ? true : false;
+        return intersects( _geometry, node->bounding_box );
     }
 
-  private:
-    Point _query_point;
-    double _radius;
+    Geometry _geometry;
 };
 
-class Overlap
-{
-  public:
-    using Tag = SpatialPredicateTag;
+using Within = Intersects<Sphere>;
+using Overlap = Intersects<Box>;
```

Comment:
This two names are pretty bad. We should get ride of them. Within as nothing to do with being inside a volume and Overlap actually applies to both. The template names are a lot more meaningful.

Reply:
I agree with you.  I am trying to come up with better names.
In the meantime these aliases will do.

Reply:
What is the Boost convention here?

Reply:
Actually do we even want a type alias here? Whats wrong with keeping them as `Intersects<T>`?

Reply:
Boost convention is to say the predicate type is "unspecified" and they use the convenience templated `intersects(geometry)` to generate the query (see [here])(http://www.boost.org/doc/libs/1_65_1/libs/geometry/doc/html/geometry/reference/spatial_indexes/group__predicates/intersects_geometry_const___.html#group__predicates_1ga78cb2ef221b951867c591ffb51b7d8c5).
We need to spell out the actual type of the queries because we make a Kokkos view of them.